aioxmpp/pep/service.py

```diff
@@ -0,0 +1,363 @@
+    """
+    :class:`PEPClient` simplifies working with PEP services.
+
+    Compared to :class:`~aioxmpp.PubSubClient` it supports automatic
+    checking for server support, a stream-lined API. It is intended to
+    make PEP things easy. If you need more fine-grained control or do
+    things which are not usually handled by the defaults when using PEP, use
+    :class:`~aioxmpp.PubSubClient` directly.
+
+    See :class:`register_pep_node` for the high-level interface for
+    claiming a PEP node and receiving event notifications.
+
+    There also is a low-level interface for claiming nodes:
+
+    .. automethod:: is_claimed
+
+    .. automethod:: claim_pep_node
+
+    Further we have a convenience method for publishing items in the client's
+    PEP service:
+
+    .. automethod:: publish
+
+    Use the :class:`aioxmpp:PubSubClient` For explicit subscription
+    and unsubscription .
+    """
+    ORDER_AFTER = [
+        aioxmpp.DiscoClient,
+        aioxmpp.DiscoServer,
+        aioxmpp.PubSubClient,
+    ]
+
+    def __init__(self, client, **kwargs):
+        super().__init__(client, **kwargs)
+        self._pubsub = self.dependencies[aioxmpp.PubSubClient]
+        self._disco_client = self.dependencies[aioxmpp.DiscoClient]
+        self._disco_server = self.dependencies[aioxmpp.DiscoServer]
+
+        self._pep_node_claims = weakref.WeakValueDictionary()
+
+    def is_claimed(self, node):
+        """
+        Return whether `node` is claimed.
+        """
+        return node in self._pep_node_claims
+
+    def claim_pep_node(self, node_namespace, *,
+                       register_feature=True, notify=False):
+        """
+        Claim node `node_namespace`.
+
+        :param node_namespace: the pubsub node whose events shall be
+            handled.
+        :param handler: the handler to install.
+        :type handler: callable, see
+            :attr:`aioxmpp.PubSubClient.on_item_publish`
+            for the arguments.
+        :param register_feature: Whether to publish the `node_namespace`
+            as feature.
+        :param notify: Whether to register the ``+notify`` feature to
+            receive notification without explicit subscription.
+        :raises RuntimeError: if a handler for `node_namespace` is already
+            set.
+        :returns: a :class:`RegisteredPEPNode` instance representing
+            the claim.
+
+        This registers `node_namespace` as feature for service discovery
+        unless ``register_feature=False`` is passed.
+        """
+        if node_namespace in self._pep_node_claims:
+            raise RuntimeError(
+                "claiming already claimed node"
+            )
+        registered_node = RegisteredPEPNode(
+            self,
+            node_namespace,
+            register_feature=register_feature,
+            notify=notify,
+        )
+
+        finalizer = weakref.finalize(
+            registered_node,
+            weakref.WeakMethod(registered_node._unregister)
```

Comment:
Shouldn’t this call ``close`` instead of ``_unregister``? Two benefits: If it has been closed, it won’t fail additionally and it will clean up the internal stuff in PEPService too.